fs/io_zenfs.cc

```diff
@@ -632,6 +633,174 @@ size_t ZonedRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
+
+    read += ret;
+    read_pos += ret;
+  }
+
+  return IOStatus::OK();
+}
+
+// This is a heavy weight function. There is going to be a high
+// traffic activity via the PCIe channel to the ZNS SSD because
+// of the read/write to zones which needs  to be issued. We need
+// some better ideas later to bring in efficiency, something like
+// "simple copy" or ideas in those lines.
+IOStatus ZenFSGCWorker::MoveValidDataToNewDestZone() {
+  std::vector<Zone*>::iterator zone_it;
+  std::vector<ZoneExtent*>::iterator ext_it;
+  std::vector<ZoneExtent*>::iterator e_it;
+  IOStatus s;
+  uint64_t r_pos;
+  uint32_t size;
+  uint32_t long_ext_size;
+  uint64_t new_start;
+  void* align_buf;
+  int dont_read = 0;
+
+  // Sort the Extent list in decreasing order.
+  std::sort(extent_list.begin(), extent_list.end(),
+            [](ZoneExtent* ext1, ZoneExtent* ext2) {
+              return ext1->length_ > ext2->length_;
+            });
+
+  // Get the size of the largest extent.
+  long_ext_size = extent_list[0]->length_;
+
+  // Allocate a aligned buffer with size of the largest extent.
+  // We have to issue pread from the source zones so we need a
+  // buffer.
+  int ret = posix_memalign(&align_buf, sysconf(_SC_PAGESIZE), long_ext_size);
+  if (ret) return IOStatus::IOError("Failed to allocate aligned memory");
+
+  zone_it = dst_zone_list.begin();
+  for (ext_it = extent_list.begin(); ext_it != extent_list.end();) {
+    ZoneExtent* ext;
+    Zone* zone_dst;
+
+    ext = *ext_it;
+    zone_dst = *zone_it;
+
+    // Set the position and length in the source zone to
+    // read the data.
+    r_pos = ext->start_;
+    size = ext->length_;
+    Slice buf((const char*)align_buf, size);
+
+    if (!dont_read) {
+      s = ReadExtent(&buf, r_pos, ext->zone_);
+      if (!s.ok()) {
+        free(align_buf);
+        return s;
+      }
+    }
+
+    // Store the new starting position for the extent
+    // which will be later made persistent.
+    new_start = zone_dst->wp_;
+
+    // Write the valid data where were read from the
+    // source zone to the destination zone.
+    s = zone_dst->Append((char*)align_buf, size);
+    if (s.ok()) {
+      // Data was written to the new zone, so the extent
+      // will have a new starting position. No need to
+      // change the length of the extent as it will be the
+      // same.
+      ext->start_ = new_start;
+
+      // The extent was moved to a new zone so change the
+      // resident zone parameter of the extent.
+      ext->zone_ = zone_dst;
+
+      // Current extent was written so now fetch the next extent.
+      ext_it++;
+      memset(align_buf, 0, long_ext_size);
+      dont_read = 0;
+      continue;
+    }
+
+    if (s == IOStatus::NoSpace()) {
+      // Data was already read before, no need to read it again
+      dont_read = 1;
+
+      // The current zone cannot fit this extent because of lack
+      // of space, so get the next zone from the dst_zone_list.
+      zone_it++;
```

Comment:
Should we consider the case that there is not enough destination zone?

Reply:
> Should we consider the case that there is not enough destination zone?

Yes, that possibility could arise. The function GetDestZoneToMoveValidData() should do that math and figure out how many zones are needed to fit all the extents(cold valid data) and called AllocateZone() for moving the valid cold data. If the desired amount of zones to fit all the cold valid data( or extents ) is not possible to allocate then it will send IOStatus error that "Not enough zones". In that case, the GC cannot progress. Maybe a slight design change can be done that we try to move only that much valid data, for which we have space(zones). Since this is a periodic background reclaim, in the next cycles, the remaining  zones could be reclaimed. Just a thought, but this could make the implementation intricate.